trieste/models/gpflow/models.py

```diff
@@ -1362,3 +1368,259 @@ def covariance_between_points(
+
+            mask = query_points_fidelity_col >= fidelity
+
+            observation_noise = tf.where(mask, fidelity_observation_noise, observation_noise)
+
+        return f_mean, f_var + observation_noise
+
+    def update(self, dataset: Dataset) -> None:
+        """
+        Update the models on their corresponding data. The data for each model is
+        extracted by splitting the observations in ``dataset`` by fidelity level.
+
+        :param dataset: The query points and observations for *all* the wrapped models.
+        """
+        check_and_extract_fidelity_query_points(
+            dataset.query_points, max_fidelity=self.num_fidelities - 1
+        )
+        dataset_per_fidelity = split_dataset_by_fidelity(dataset, self.num_fidelities)
+        for fidelity, dataset_for_fidelity in enumerate(dataset_per_fidelity):
+            if fidelity == 0:
+                self.lowest_fidelity_signal_model.update(dataset_for_fidelity)
+            else:
+                # Make query points but with final column corresponding to
+                # fidelity we wish to predict at
+                self.fidelity_residual_models[fidelity].update(
+                    Dataset(
+                        dataset_for_fidelity.query_points,
+                        self._calculate_residual(dataset_for_fidelity, fidelity),
+                    )
+                )
+
+    def optimize(self, dataset: Dataset) -> None:
+        """
+        Optimize all the models on their corresponding data. The data for each model is
+        extracted by splitting the observations in ``dataset``  by fidelity level.
+        Note that we have to code up a custom loss function when optimizing our residual
+        model, so that we can include the correlation parameter as an optimisation variable.
+
+        :param dataset: The query points and observations for *all* the wrapped models.
+        """
+        check_and_extract_fidelity_query_points(
+            dataset.query_points, max_fidelity=self.num_fidelities - 1
+        )
+        dataset_per_fidelity = split_dataset_by_fidelity(dataset, self.num_fidelities)
+
+        for fidelity, dataset_for_fidelity in enumerate(dataset_per_fidelity):
+            if fidelity == 0:
+                self.lowest_fidelity_signal_model.optimize(dataset_for_fidelity)
+            else:
+                gpf_residual_model = self.fidelity_residual_models[fidelity].model
+
+                fidelity_observations = dataset_for_fidelity.observations
+                fidelity_query_points = dataset_for_fidelity.query_points
+                prev_fidelity_query_points = convert_query_points_for_fidelity(
+                    dataset_for_fidelity.query_points, fidelity - 1
+                )
+                predictions_from_lower_fidelity = self.predict(prev_fidelity_query_points)[0]
+
+                def loss() -> TensorType:  # hardcoded log liklihood calculation for residual model
```

Comment:
I haven't checked but I'm assuming this is mimicking the `gpflow` version as much as possible? there might be subtleties we could miss otherwise that could affect performance

Reply:
I believe it is, yes